<launch>
  <arg name="sim" default="true" />

  <group if="$(arg sim)" >
    <node pkg="giskard_examples" type="whole_body_controller" name="whole_body_controller" output="screen">
      <param name="controller_descriptions/cart_cart" 
        textfile="$(find giskard_examples)/controller_specs/pr2_cart_cart_control.yaml" />
      <param name="controller_descriptions/joint_joint" 
        textfile="$(find giskard_examples)/controller_specs/pr2_joint_joint_control.yaml" />
      <param name="controller_descriptions/joint_cart" 
        textfile="$(find giskard_examples)/controller_specs/pr2_joint_cart_control.yaml" />
      <param name="controller_descriptions/cart_joint" 
        textfile="$(find giskard_examples)/controller_specs/pr2_cart_joint_control.yaml" />
      <rosparam command="load"
        file="$(find giskard_examples)/config/pr2_whole_body_controller.yaml" />
      <remap from="~joint_states" to="/joint_states" />
    </node>
  </group>

  <group unless="$(arg sim)" >
    <node pkg="giskard_examples" type="whole_body_controller" name="whole_body_controller" output="screen">
      <param name="controller_descriptions/cart_cart" 
        textfile="$(find giskard_examples)/controller_specs/iai_pr2_cart_cart_control.yaml" />
      <param name="controller_descriptions/joint_joint" 
        textfile="$(find giskard_examples)/controller_specs/pr2_joint_joint_control.yaml" />
      <param name="controller_descriptions/joint_cart" 
        textfile="$(find giskard_examples)/controller_specs/pr2_joint_cart_control.yaml" />
      <param name="controller_descriptions/cart_joint" 
        textfile="$(find giskard_examples)/controller_specs/iai_pr2_cart_joint_control.yaml" />
      <rosparam command="load"
        file="$(find giskard_examples)/config/pr2_whole_body_controller.yaml" />
      <remap from="~joint_states" to="/joint_states" />
    </node>
  </group>

  <node pkg="giskard_examples" type="float_array_separator" name="velocity_cmd_separator" output="screen" if="($arg sim)" >
    <rosparam param="joint_names"
      file="$(find giskard_examples)/config/pr2_upper_body.yaml" />
    <remap from="~input_array" to="/whole_body_controller/velocity_cmd" />
  </node>

  <node pkg="giskard_examples" type="pr2_interface_node" name="pr2_interface_right_arm"
        output="screen" unless="($arg sim)" >
    <remap from="~joint_states" to="/joint_states" />
    <remap from="~position_cmd" to="/r_arm_controller/command" />
    <remap from="~velocity_cmd" to="/whole_body_controller/velocity_cmd" />
    <rosparam command="load" file="$(find giskard_examples)/config/pr2_right_arm_joint_names.yaml" />
    <param name="period" value="0.01" />
  </node>

  <node pkg="giskard_examples" type="pr2_interface_node" name="pr2_interface_left_arm"
        output="screen" unless="($arg sim)" >
    <remap from="~joint_states" to="/joint_states" />
    <remap from="~position_cmd" to="/l_arm_controller/command" />
    <remap from="~velocity_cmd" to="/whole_body_controller/velocity_cmd" />
    <rosparam command="load" file="$(find giskard_examples)/config/pr2_left_arm_joint_names.yaml" />
    <param name="period" value="0.01" />
  </node>

  <node pkg="giskard_examples" type="controller_action_server" name="controller_action_server" 
        output="screen">
    <remap from="~command" to="whole_body_controller/goal" />
    <remap from="~feedback" to="whole_body_controller/feedback" />
    <rosparam command="load"
      file="$(find giskard_examples)/config/pr2_controller_action_server.yaml" />
  </node>
</launch>
